sycl/include/CL/sycl/reduction.hpp

```diff
@@ -100,5 +100,57 @@ reduction(T *Var, const T &Identity, BinaryOperation Combiner,
   return {Var, Identity, Combiner, InitializeToIdentity};
 }
 
+#if __cplusplus >= 201703L
```

Comment:
> 1) **Is there a better way to define the specialization of `reducer` to avoid potential ABI breakage in the future?**
  This implementation works by privatizing the entire span on each work-item, which will fail for sufficiently large spans.  Eventually we'll need to choose a different `reducer` implementation based on the `Extent` parameter.  Should `reducer` and `reducer_element` be SFINAE'd away for certain `Extent` values, even in this first implementation?

Reply:
Do you have a specific scenario in mind where this would break ABI? Seems to me like this is all in headers and since the extent is a template parameter I am not sure if user-code incompatibility should be much of a concern.
That said, if we know what `Extent` we need to stop at, I don't mind an `enable_if` masking the definition out or alternatively a `static_assert` telling the user that the given `Extent` isn't yet supported. Is the `Extent` limit resource-constrained? If so, I think we may as well let the user do as wild reductions as their hardware allows until we have a good alternative path to take based on some heuristic.

Reply:
It makes sense to me. Quite frankly I expected we would eventually reach a point where we would need to expose different reduction strategies to the user and what you suggest is pretty much what I expected.

How exactly we would want to approach that though is a good question. It sounds like this choice would be user-facing and as such we would probably need an extension. That said, it seems like it would be a great excuse to bring compile-time properties to `reduction`.

Reply:
I've restructured things to avoid code duplication in https://github.com/intel/llvm/pull/6019/commits/111ea2f6c11ac9e98a16fb01958eb44a04a14d0a.  While I was doing it I've added a few new template arguments that we can use for specialization.

I'm using an algorithm class to store all the implementation-specific stuff because it gives us a bit more flexibility than an enum, but I think once we've got a few more of these we should definitely think about exposing a few compile-time properties.  We've already had a request for a few different properties (see https://github.com/intel/llvm/issues/1621).  A good test of this new class structure would be whether we can move the existing `SYCL_REDUCTION_DETERMINISTIC` into something like a `deterministic_reduction_algorithm` class and just plug it in.

I'm hoping that the combination of `Dims`, `Extent` and `Algorithm` will be enough to do the filtering for the parameter pack. 🤞